ios/app/AudioMode.m

```diff
@@ -0,0 +1,130 @@
+#import "AudioMode.h"
+#import "RCTLog.h"
+
+
+@implementation AudioMode
+
+RCT_EXPORT_MODULE();
+
+typedef enum {
+    kAudioModeDefault,
+    kAudioModeAudioCall,
+    kAudioModeVideoCall
+} JitsiMeetAudioMode;
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        _initialized = NO;
+        _category = nil;
+        _mode = nil;
+        _session = [AVAudioSession sharedInstance];
+    }
+    return self;
+}
+
+- (dispatch_queue_t)methodQueue
+{
+    // Make sure all our methods run in the main thread.  The route change
+    // notification runs there so this will make sure it will only be fired
+    // after our changes have been applied (when we cause them, that is).
+    return dispatch_get_main_queue();
+}
+
+- (void)routeChanged:(NSNotification*)notification {
+    NSDictionary *dict = notification.userInfo;
+    NSInteger reason = [[dict valueForKey:AVAudioSessionRouteChangeReasonKey]
+                        integerValue];
+    switch (reason) {
+    case AVAudioSessionRouteChangeReasonCategoryChange: {
+        // The category has changed, check if it's the one we want and adjust
+        // as needed.
+        BOOL success;
+        NSError *error;
+
+        if (_session.category != _category) {
+            success = [_session setCategory: _category error: &error];
+            if (!success || error) {
+                RCTLogInfo(@"Error overriding the desired session category");
+            }
+        }
+
+        if (_session.mode != _mode) {
+            success = [_session setMode: _mode error: &error];
+            if (!success || error) {
+                RCTLogInfo(@"Error overriding the desired session mode");
+            }
+        }
```

Comment:
I'm concerned about `routeChange:` and 'setMode:resolve:reject:' having very similar source code that's core to this feature: `setCategory:error:` and `setMode:error:`. I think a common method invoked by the two makes me feel more comfortable.